### DISCUSSION TITLE
Skypack, Preact, JSX Intrinsic Elements.

### DISCUSSION THREAD

**Discussion by tbremer** [denoland/deno#9855] [Q&A]
I have read through a couple of issues related to this, but not one that directly gives a solution.

I have a project where I was wanting to use Preact, my `deps.ts` looks like this:

```typescript
import * as preactRenderToString from "https://cdn.skypack.dev/preact-render-to-string?dts";
const { render } = preactRenderToString;

import * as preact from "https://cdn.skypack.dev/preact?dts";
export { preact, render };
```

As I was using Preact, which has a different jsx factory, I had to use a custom `tsconfig.json`. It looks like this:

```json
{
    "compilerOptions": {
        "lib": ["esnext", "dom", "deno.ns"],
        "jsx": "react",
        "jsxFactory": "h",
        "jsxFragmentFactory": "Fragment"
    }
}
```

Skypack resolves modules with type definitions using their `?dts` query paramter. If I reload the cache I can see the `*.d.ts` files being fetched:

```
$ deno cache --reload --config tsconfig.json deps.ts 
Download https://cdn.skypack.dev/preact-render-to-string?dts
Download https://cdn.skypack.dev/preact?dts
Download https://cdn.skypack.dev/-/preact@v10.5.13-YBEbAK5M5p0s1c9aovEb/dist=es2020,mode=imports/optimized/preact.js
Download https://cdn.skypack.dev/-/preact@v10.5.13-YBEbAK5M5p0s1c9aovEb/dist=es2020,mode=types/src/index.d.ts
Download https://cdn.skypack.dev/-/preact@v10.5.13-YBEbAK5M5p0s1c9aovEb/dist=es2020,mode=types/src/jsx.d.ts
Download https://cdn.skypack.dev/-/preact-render-to-string@v5.1.16-OpLkEM3eRx574UMXNSoo/dist=es2020,mode=imports/optimized/preact-render-to-string.js
Download https://cdn.skypack.dev/-/preact-render-to-string@v5.1.16-OpLkEM3eRx574UMXNSoo/dist=es2020,mode=types/src/index.d.ts
Download https://cdn.skypack.dev/-/preact@v10.5.13-V3GjhSMJg71SAoQzY34u/dist=es2020,mode=imports/optimized/preact.js
Download https://cdn.skypack.dev/-/preact@v10.5.13-V3GjhSMJg71SAoQzY34u/dist=es2020,mode=types/index.d.ts
Download https://cdn.skypack.dev/-/preact@v10.5.13-V3GjhSMJg71SAoQzY34u/dist=es2020,mode=types/src/index.d.ts
Download https://cdn.skypack.dev/-/preact@v10.5.13-V3GjhSMJg71SAoQzY34u/dist=es2020,mode=types/src/jsx.d.ts
Check file:///app/deps.ts
```

Notably, the last file downloaded `https://cdn.skypack.dev/-/preact@v10.5.13-V3GjhSMJg71SAoQzY34u/dist=es2020,mode=types/src/jsx.d.ts` contains the `JSX` namespace that I am having an issue with.

Whenever I actually use jsx within the code:
```typescript
import { render, preact } from "./deps.ts";
const { h } = preact;

console.log(render(<h1>hello, world!</h1>));
```

I am returned an error about the JSX namespace:

```
TS7026 [ERROR]: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists.
  return <h1>Hello, world!</h1>;
```

The Deno docs (https://deno.land/manual@v1.8.1/typescript/types#type-declaration-semantics) aren't clear, outside of suggesting Skypack.

Is there a solution outside of hosting the module directly like @lucacasonato did for dext.ts?

**Comment by lucacasonato** [comment]
https://x.lcas.dev/preact@10.5.11/mod.js uses different typings. The typings that are served by Skypack do not work in Deno (they rely on namespace augmentation).

**Reply by tbremer** [reply]
Thanks for the response! I looked over your types, but only briefly and did not realize there was a difference.

**Comment by tbremer** [comment]
FWIW, I opened an issue with Preact to see if we can get this resolved.

https://github.com/preactjs/preact/issues/3075

**Reply by marvinhagemeister** [reply]
Preact maintainer here.

We're thinking of moving back to defining JSX types in the global namespace due to most modern bundlers injecting different jsx runtimes depending on the build time. Due to the automatic jsx factory injection `h` or `createElement` is never imported from TypeScript's point of view and the namespaced JSX types are not found. Promoting them to the global scope would fix it in those scenarios.

I'm not too familiar with how deno works and leverages the TS compiler. Would promoting JSX to the global scope help here?

**Reply by tbremer** [reply]
Hopefully a Deno maintainer can chime in and fill us in with the particulars as I am also not terribly familiar with Deno's interaction with the compiler, but as far as I know Deno does not allow/put anything in the global namespace. Saying: _It is unpredictable how Deno will handle ambient/global declarations._

https://deno.land/manual@v1.8.1/typescript/types#type-declaration-semantics

**Reply by tbremer** [reply]
I would _guess_ that Global to a module and Global to the environment are different in this case.

**Reply by aadamsx** [reply]
@tbremer I was just today researching/considering using Preact for my next Deno project and came across this post. Are you unable to actually use Preact with Deno? From your starting post it seems you can't even do a simple render in Deno.

Maybe Nanojsx is a better solution at the moment (it supports Deno out of the box)?  https://nanojsx.github.io/

**Reply by kitsonk** [reply]
>  but as far as I know Deno does not allow/put anything in the global namespace

No, it does...  It is just that ambient declarations are unpredictable, because Deno may not seem them, because it does not automatically acquire `.d.ts` files and treats everything as a module...  If it is a UMD declaration that does a `declare global` then it should be fine.

@marvinhagemeister I suspect @lucacasonato or myself would be glad to try to help.  I will try to find a time to take a look and suggest something that would work for Deno and hopefully everyone else.

**Reply by marvinhagemeister** [reply]
@kitsonk That's great to hear! I'd really love to make the Preact + Deno experience as smooth as possible. I'm unsure where to start debugging this. With node I'll usually just mess with files in `node_modules` directly and fire up the tsc compiler again.

Fun fact: We're currently working on improving our server rendering performance and one of the test cases is coincidently the deno docs as they are a good benchmark for that!

**Reply by tbremer** [reply]
> No, it does... It is just that ambient declarations are unpredictable, because Deno may not seem them, because it does not automatically acquire `.d.ts` files and treats everything as a module... If it is a UMD declaration that does a `declare global` then it should be fine.

thanks for clarifying!

**Reply by tbremer** [reply]
@marvinhagemeister, @kitsonk please let me know if there is some way I can help!
